### PR TITLE
prov/util: add debugging in ofi_mr_map_verify

### PR DESCRIPTION
**Pull request by tschuett**
state the access violations

Signed-off-by: Thorsten Schütt <schuett@gmail.com>

### CODE REVIEW

File: prov/util/src/util_mr_map.c [hunk 1]
```diff
@@ -108,22 +108,32 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 	void *addr;
 
 	node = ofi_rbmap_find(map->rbtree, &key);
-	if (!node)
-		return -FI_EINVAL;
+	if (!node) {
+                FI_DBG(map->prov, FI_LOG_MR, "verify_addr: unknown key: %" PRIu64 "\n",
+                  key);
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:113]: Please check line length for all new calls (several look long) and fix alignment.  All of these should probably be FI_WARN, as they are unexpected errors.

File: prov/util/src/util_mr_map.c [hunk 2]
```diff
@@ -108,22 +108,32 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
...
 		return -FI_EACCES;
 	}
 
 	addr = (void *) (*io_addr + (uintptr_t) attr->offset);
 
-	if ((addr < attr->mr_iov[0].iov_base) ||
-	    (((char *) addr + len) > ((char *) attr->mr_iov[0].iov_base +
-			    	      attr->mr_iov[0].iov_len))) {
+	if (addr < attr->mr_iov[0].iov_base) {
+                FI_DBG(map->prov, FI_LOG_MR, "verify_addr: access hit before mr range\n");
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:130]: I'd merge the checks back together and modify the text.  Something similar to "target region (0x%zx - 0x%zx) out of registered range (0x%zx - 0x%zx)", addr, addr + len, iov_base, iov_base + iov_len)

File: prov/util/src/util_mr_map.c [hunk 3]
```diff
@@ -108,22 +108,32 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
...
+        }
 
 	attr = node->data;
 	assert(attr);
 
 	if ((access & attr->access) != access) {
-		FI_DBG(map->prov, FI_LOG_MR, "verify_addr: invalid access\n");
+                FI_DBG(map->prov, FI_LOG_MR, "verify_addr: invalid access: %s and %s\n",
+                  fi_tostr(&attr->access, FI_TYPE_MR_MODE),
+                  fi_tostr(&access, FI_TYPE_MR_MODE));
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:123]: fi_tostr() uses an internal character array.  Calling it twice will replace the contents of the array.  Either use 2 separate prints, one per fi_tostr() call, print the flags in hex, or use fi_tostr_r() and pass in buffers.  The easiest but most useful is 2 prints.

The text here should indicate "permitted access: xxx, requested access: yyy", to make it clear which value is which.

**Review by shefty — CHANGES_REQUESTED** [review]

**Comment by j-xiong** [comment]
Please fix the title:  "of_" --> "ofi_". Also don't create a new commit to fix a previous commit. Use force push to update the commit directly. This project is not configured to squash upon merge.

**Comment by tschuett** [comment]
The last commit was a force push. Do you want me to open a new PR?

**Comment by j-xiong** [comment]
@tschuett I am seeing 5 commits now (there should be only one). There is no need to create a new PR to fix this. Do you see 5 new commits on your local machine? The proper way to fix a single commit is:

1. change the code
2. `git commit -a --amend`
3. `git push -f`

Use git rebase or reset fi your local branch already contains unwanted extra commits.

**Comment by tschuett** [comment]
Thanks. But I cannot eliminate the earlier commits.

**Comment by j-xiong** [comment]
@tschuett Do this:
```
git reset HEAD^^^^
git commit -a --amend
git push -f
```

**Comment by j-xiong** [comment]
Looks clean now. Please also update the title of the commit message -- remove the brackets and add a colon.

**Comment by tschuett** [comment]
Thanks!

File: prov/util/src/util_mr_map.c [hunk 4]
```diff
@@ -108,22 +108,37 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 	void *addr;
 
 	node = ofi_rbmap_find(map->rbtree, &key);
-	if (!node)
-		return -FI_EINVAL;
+	if (!node) {
+                FI_WARN(map->prov, FI_LOG_MR,
+                    "verify_addr: unknown key: %" PRIu64 "\n", key);
```
Comment (j-xiong) [inline prov/util/src/util_mr_map.c:113]: Please align with the "m" in the previous line. Same for other places.

Comment (tschuett) [reply prov/util/src/util_mr_map.c]: Fixed.

**Review by j-xiong — APPROVED** [review]

**Comment by tschuett** [comment]
Thank you for your patience. This is my first PR in this repository.

**Comment by j-xiong** [comment]
@tschuett Welcome and thanks for your contribution!

File: prov/util/src/util_mr_map.c [hunk 5]
```diff
@@ -108,22 +108,37 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
...
+                FI_WARN(map->prov, FI_LOG_MR,
+                        "verify_addr: invalid access: requested %s\n",
+                        fi_tostr(&access, FI_TYPE_MR_MODE));
 		return -FI_EACCES;
 	}
 
 	addr = (void *) (*io_addr + (uintptr_t) attr->offset);
 
-	if ((addr < attr->mr_iov[0].iov_base) ||
+        if ((addr < attr->mr_iov[0].iov_base) ||
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:132]: nit: You indented with spaces instead of a tab here.  Please revert

File: prov/util/src/util_mr_map.c [hunk 6]
```diff
@@ -108,22 +108,37 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 	void *addr;
 
 	node = ofi_rbmap_find(map->rbtree, &key);
-	if (!node)
-		return -FI_EINVAL;
+	if (!node) {
+                FI_WARN(map->prov, FI_LOG_MR,
+                        "verify_addr: unknown key: %" PRIu64 "\n", key);
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:113]: Please remove the 'verify_addr' from each of the prints.  The log message will include the real function name, so there's no need for a partial name that doesn't match.

File: prov/util/src/util_mr_map.c [hunk 7]
```diff
@@ -108,22 +108,37 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
...
 
 	addr = (void *) (*io_addr + (uintptr_t) attr->offset);
 
-	if ((addr < attr->mr_iov[0].iov_base) ||
+        if ((addr < attr->mr_iov[0].iov_base) ||
 	    (((char *) addr + len) > ((char *) attr->mr_iov[0].iov_base +
 			    	      attr->mr_iov[0].iov_len))) {
+                FI_WARN(map->prov, FI_LOG_MR,
+                        "verify_addr: target region (0x%zx - 0x%zx) "
+                        "out of registered range (0x%zx - 0x%zx)\n",
```
Comment (shefty) [inline prov/util/src/util_mr_map.c:137]: I messed up in my prior comment.  I didn't realize addr was a pointer.  You can just use %p to print all of the addresses, which lets you remove the casts to (uintptr_t) below.

File: prov/util/src/util_mr_map.c (hunk 6 above)
Comment (tschuett) [reply prov/util/src/util_mr_map.c]: Fixed. That was prior art.

File: prov/util/src/util_mr_map.c (hunk 7 above)
Comment (tschuett) [reply prov/util/src/util_mr_map.c]: Fixed.

File: prov/util/src/util_mr_map.c (hunk 5 above)
Comment (tschuett) [reply prov/util/src/util_mr_map.c]: Fixed. Editor issues.

**Review by shefty — APPROVED** [review]
thanks! - just waiting on CI results

**Comment by tschuett** [comment]
Intel CI is unhappy.

**Comment by shefty** [comment]
Intel CI is always unhappy.  It's fed nothing but sweet tarts.  Problem isn't related to this PR.

**Pull request merged**